main.py

```diff
@@ -2,23 +2,90 @@
 #
 # Simple manager prototype for xqemu
 #
-from PyQt5.QtWidgets import QApplication, QDialog, QFileDialog, QMainWindow, QMessageBox
+from PyQt5.QtWidgets import QApplication, QDialog, QFileDialog, QMainWindow, QMessageBox, QTableWidgetItem, QAbstractItemView, QHeaderView
 from PyQt5.uic import loadUiType
 from PyQt5 import QtCore, QtGui
 from qmp import QEMUMonitorProtocol
+from glob import glob1
 import sys
 import os, os.path
 import json
 import subprocess
 import time
 import platform
+import csv
+import sys
 
 SETTINGS_FILE = './settings.json'
+LIBRARY_CACHE = './library.cache'
 
 # Load UI files
 settings_class, _ = loadUiType('settings.ui')
 mainwindow_class, _ = loadUiType('mainwindow.ui')
 
+class DiscImageFormat(object):
+	XISO = 'XISO'
+	ISO9660 = 'ISO9660'
+	Unknown = 'Unknown format'
+
+def read_from_offset(file, offset, length):
+	file.seek(offset)
+	return file.read(length).decode('unicode-escape')
+
+def getDiscFileFormat(filepath):
+	file = open(filepath, 'rb')
+
+	#XISO/XGD
+	if read_from_offset(file, 0x10000, 20) == 'MICROSOFT*XBOX*MEDIA': return DiscImageFormat.XISO
+
+	# ISO 9660
+	for offset in [0x9001, 0x8801, 0x8001]:
+		if read_from_offset(file, offset, 5) == 'CD001': return DiscImageFormat.ISO9660
+
+	return DiscImageFormat.Unknown
+
+class CacheManager(object):
+	def __init__(self):
+		self.reset()
+
+	def reset(self):
+		self.cache = []
+
+	def save(self):
+		# Some python version stuff to get rid of empty lines in the cache
+		if sys.version_info[0] == 3:
+			cachefile = open(LIBRARY_CACHE, 'w', newline='')
+		else:
+			cachefile = open(LIBRARY_CACHE, 'wb')
```

Comment:
This code block feels hacky. Can you provide an example of the bad "empty lines"? Also why CSV if it has these problems? Are other formats also affected?

Reply:
This is python related cause their csv writer has some weird new line logic on windows.
JSON imo is a bit bloated cause it adds "too many" useless characters when we can just have a few commas.
This would only affect really big game libraries tho.

Reply:
Who's going to hand edit this though? Having it human readable is mostly a debug feature.
And we already use json elsewhere, so it makes sense to use it.

Reply:
I can't stress @JayFoxRox's point enough - CSV is an awful file format, and should never be used for anything:

- Line endings are part of the syntax, and are OS-dependent (CRLF on Windows, LF on unix - seems you already hit issues with that)
- The delimiter is region and encoding-dependent. For example, german Excel uses a semicolon by default, while english Excel uses a comma, unless it detects that it's a Unicode file, then it uses tab.
- The encoding is not specified (just imagine the horrors of the poor soul that decides to edit a random CSV-file with notepad or some other ignorant editor)

JSON is not perfect, but for machine readability, it's a much better format, and with a proper editor (i.e. one that supports folding) it's much more readable for humans, too.